python/task_manager/background_task_manager.py

```diff
@@ -1,47 +1,142 @@
 # Copyright (c) 2015 Shotgun Software Inc.
-# 
+#
 # CONFIDENTIAL AND PROPRIETARY
-# 
-# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit 
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
 # Source Code License included in this distribution package. See LICENSE.
-# By accessing, using, copying or modifying this work you indicate your 
-# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights 
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
 # not expressly granted therein are reserved by Shotgun Software Inc.
 
-import traceback
+"""
+Background task manager.
+"""
 
+import Queue
 import sgtk
 from sgtk.platform.qt import QtCore
 from sgtk import TankError
 
 from .background_task import BackgroundTask
 from .worker_thread import WorkerThread
 
+
 class BackgroundTaskManager(QtCore.QObject):
     """
     Main task manager class. Manages a queue of tasks running them asynchronously through
     a pool of worker threads.
 
-    Note that the BackgroundTaskManager class itself is reentrant but not thread-safe so its methods should only 
+    Note that the BackgroundTaskManager class itself is reentrant but not thread-safe so its methods should only
     be called from the thread it is created in. Typically this would be the main thread of the application.
-    
-    
+
+
     :signal task_completed(uid, group, result): Emitted when a task has been completed.
-        The ``uid`` parameter holds the unique id associated with the task, 
-        the ``group`` is the group that the task is associated with and 
-        the ``result`` is the data returned by the task. 
-    
+        The ``uid`` parameter holds the unique id associated with the task,
+        the ``group`` is the group that the task is associated with and
+        the ``result`` is the data returned by the task.
+
     :signal task_failed(uid, group, message, traceback_str): Emitted when a task fails for some reason.
-        The ``uid`` parameter holds the unique id associated with the task, 
-        the ``group`` is the group that the task is associated with, 
+        The ``uid`` parameter holds the unique id associated with the task,
+        the ``group`` is the group that the task is associated with,
         the ``message`` is a short error message and the ``traceback_str``
         holds a full traceback.
-        
+
     :signal task_group_finished(group): Emitted when all tasks in a group have finished.
         The ``group`` is the group that has completed.
 
     """
 
+    class _ResultsPoller(QtCore.QObject):
+        """
+        Polls a queue which holds the results posted from the worker threads.
```

Comment:
In a chat with JF today we were talking about doing the select in a subthread and then scheduling the task to run on the right thread without signaling.  To avoid having to poll. We'll see if it is doable.